experimental/DoubleAndHyperComplexes/src/Morphisms/cartan_eilenberg_resolutions.jl

```diff
@@ -0,0 +1,59 @@
+mutable struct MapLifter{MorphismType} <: HyperComplexMorphismFactory{MorphismType}
+  dom::AbsHyperComplex
+  cod::AbsHyperComplex
+  orig_map::Map
+  start_index::Int
+  offset::Int
+  check::Bool
+
+  function MapLifter(::Type{MorphismType}, dom::AbsHyperComplex, cod::AbsHyperComplex, phi::Map; 
+      start_index::Int=0, offset::Int=0, check::Bool=true
+    ) where {MorphismType}
+    @assert dim(dom) == 1 && dim(cod) == 1 "lifting of maps is only implemented in dimension one"
+    @assert (is_chain_complex(dom) && is_chain_complex(cod)) || (is_cochain_complex(dom) && is_cochain_complex(cod))
+    @assert domain(phi) === dom[start_index]
+    @assert codomain(phi) === cod[start_index + offset]
+    return new{MorphismType}(dom, cod, phi, start_index, offset)
+  end
+end
+
+is_chain_complex(c::AbsHyperComplex) = (dim(c) == 1 ? direction(c, 1) == (:chain) : error("complex is not one-dimensional"))
+is_cochain_complex(c::AbsHyperComplex) = (dim(c) == 1 ? direction(c, 1) == (:cochain) : error("complex is not one-dimensional"))
+
+
+function (l::MapLifter{T})(Phi::AbsHyperComplexMorphism, I::Tuple) where {T<:ModuleFPHom}
+  i = first(I)
+  i == l.start_index && return l.orig_map
+
+  inc = 1 
+  if is_chain_complex(domain(Phi))
+    inc = -1
+  end
+  psi = Phi[Tuple(i+inc)]
+  dom_map = map(domain(Phi), 1, I)
+  cod_map = map(codomain(Phi), 1, Tuple(i + l.offset))
+  @assert codomain(dom_map) === domain(psi)
+  @assert codomain(cod_map) === codomain(psi)
+
+  x = gens(domain(dom_map))
+  y = dom_map.(x)
+  z = psi.(y)
+  w = [preimage(cod_map, a) for a in z]
```

Comment:
One the primary benefits of broadcast operators is that you can join them to avoid redundant allocations by fusing operations. For example:
```julia
julia> f(v) = cos.(sin.(v));

julia> g(v) = begin w = sin.(v) ; cos.(w) end;

julia> @btime f(v);
  52.063 ns (1 allocation: 80 bytes)

julia> @btime g(v);
  67.810 ns (2 allocations: 160 bytes)
```
So that "naive" version allocates two vectors, the one using chained broadcasting just one.

That would suggest at least doing
```suggestion
  z = psi.(dom_map.(x))
  w = [preimage(cod_map, a) for a in z]
```
But broadcasting also supports handling "scalars" differently to allow using it for multiple arguments as well.

I didn't test it, but maybe even this works:
```suggestion
  w = preimage.(cod_map, psi.(dom_map.(x)))
```

Now of course this is a fairly minor operation, but I am trying to raise awareness of this nevertheless, as all these tiny allocations everywhere in our code base add up -- one of the primary reasons our test suite uses many *gigabytes* for RAM for many computations for which it really shouldn't (the other of course is memory used to compile code and store the result).

Reply:
Yes, I get it. Thanks for the hints.